### ISSUE TITLE
Rpc finality change

### ISSUE THREAD

**Issue by bowenwang1996** [near/near-api-js#261]
We changed query rpcs to by default query the state as of the *last final block*, not the last known block in nearcore https://github.com/nearprotocol/nearcore/pull/2186. However, I failed to realize the consequences of this change for applayer and this change, when deployed, will potentially lead to a nonfunctional wallet: wallet creates some account and then query it the account will be shown as nonexistent because the transaction hasn't been finalized yet. On the other hand, in some other places (wallet balance for example), we probably do want to display only the finalized state. It's not immediately clear to me what the best approach here is. @vgrichina @kcole16

**Comment by bowenwang1996** [comment]
Update: I was slightly wrong. I made the change backward compatible so that the behavior is the same on the applayer side. But we should still figure out how we should leverage finality.

**Comment by potatodepaulo** [comment]
@bowenwang1996 what needs to change in near-api-js to reflect this? Would we need to wait on the transaction to finalize?

**Comment by bowenwang1996** [comment]
Actually I believe on wallet/explorer side we handle it by showing finality as part of the status. I guess there isn't much left to be done, except that maybe we should use `final` as the default parameter in rpcs. @kcole16 is that correct?

**Comment by potatodepaulo** [comment]
This is trivial to add after this is finished: https://github.com/near/near-api-js/issues/338.

@chadoh thoughts on the design for parameters?

**Comment by potatodepaulo** [comment]
Acceptance criteria:
* Implement this in the near-shell: https://docs.near.org/docs/api/rpc-params#using-near-final-finality 
* Blocked by change in named parameters. See:

**Comment by kcole16** [comment]
@bowenwang1996 For implementing #344, do we need to add `finality: 'final'` to the RPC calls?

**Comment by bowenwang1996** [comment]
@kcole16 not really. They are two parallel things. Also for transaction status query we don't return finality yet.

**Comment by kcole16** [comment]
So if we have a TX with a status, do we then need to query the block to see if it is finalized? Or can we infer TX is in a finalized if we receive a status?

**Comment by bowenwang1996** [comment]
> So if we have a TX with a status, do we then need to query the block to see if it is finalized?

That's right. I think we have a policy that if something information can be inferred we don't return it from rpc. In this case, you can infer whether the transaction is finalized by querying the block to see whether the block is finalized.

**Comment by kcole16** [comment]
In this case, that is a pretty inconvenient inference. This means we (and any other devs) need to query both the TX and the block in order to determine finality.

**Comment by bowenwang1996** [comment]
@frol what do you think?

**Comment by vgrichina** [comment]
@bowenwang1996 as far as I know there is one place in wallet/most apps where it is necessary to see finalized state – it is transaction status. It is useful to know when given transaction is final. 

Funnily it's also exactly one place in nearcore where finality isn't exposed.

For all other cases like getting current state of account (e.g. nonce to use for a given access key or balance to check before allowing user to spend money in UI) it makes far more sense to use the latest state available to RPC node. 

If it's demonstrated that rollbacks happen relatively frequently (say for more than 1% of transactions) – then we might need to reconsider.

> That's right. I think we have a policy that if something information can be inferred we don't return it from rpc. In this case, you can infer whether the transaction is finalized by querying the block to see whether the block is finalized.

1) When you need to make additional query – it actually means that information cannot be inferred. Every additional query per user's action is UX (latency) and DevX (complexity) degradation.
2) Policy sounds very counter-productive in general. What RPC returns should better be driven by needs of the RPC clients (i.e. minimize number of requests and latency to achieve common tasks).

> IMO, we should return finality (or accept it as a parameter) with TX status.

@kcole16 agree 100%. We need finality _returned_ with TX status. Being able to query by finality would be less convenient. Ability to specify finality for other RPCs doesn't seem very useful for real life use cases, as in app business logic you have to use latest values optimistically in overwhelming majority of cases (except when user needs to make sure given transaction went through – which is when we need to get finality with TX status).

**Comment by chadoh** [comment]
This seems like it has turned into a request for a change in nearcore, so I’ve assigned this back to you, @bowenwang1996. If this issue doesn’t seem like the best place to keep track of the new request, feel free to create a new issue and close this one.

**Comment by volovyk-s** [comment]
@bowenwang1996 @vgrichina is this issue still relevant?
Looks like we are using `finality` next way:
JSON RPC Docs: https://docs.near.org/docs/develop/front-end/rpc#using-finality-param
 In `near-api-js` - > provider -> block method: https://github.com/near/near-api-js/blob/0aefdb01a151f7361463f3ff65c77dbfeee83200/src/providers/json-rpc-provider.ts#L90
 In 'near-api-js' -> account -> viewState: https://github.com/near/near-api-js/blob/7564facf3eb9c3a31470e0c2962791ceab02a916/src/account.ts#L395

**Comment by frol** [comment]
This issue is not relevant for near-api-js anymore since it already provides full access to the JSON RPC parameters.
